setup.py

```diff
@@ -1,48 +1,86 @@
 from __future__ import absolute_import, division, print_function
 
-from setuptools import setup, find_packages, Extension
+from distutils.command.sdist import sdist as _sdist
+import glob
 import os
+import sys
+from setuptools import setup, find_packages, Extension
+from setuptools.command.build_ext import build_ext as _build_ext
 
-import numpy as np
-
-from Cython.Build import cythonize
+PACKAGE_NAME = 'stratify'
+PACKAGE_DIR = os.path.abspath(os.path.dirname(__file__))
 
 
-NAME = 'stratify'
-DIR = os.path.abspath(os.path.dirname(__file__))
+try:
+    # Detect if Cython is available. Where it is, use it, otherwise fall back
+    # to C source included with source distribution.
+    from Cython.Build import cythonize
+except ImportError:
+    cythonize = None
 
-extension_kwargs = {'include_dirs': [np.get_include()]}
+source_suffix = '.pyx' if cythonize else '.c'
+extension_kwargs = {}
+cython_directives = {'binding': True}
 cython_coverage_enabled = os.environ.get('CYTHON_COVERAGE', None)
-if cython_coverage_enabled:
+if cythonize and cython_coverage_enabled:
     extension_kwargs.update({'define_macros': [('CYTHON_TRACE_NOGIL', '1')]})
+    cython_directives.update({'linetrace': True})
+
+extensions = []
+for source_file in glob.glob('{}/*{}'.format(PACKAGE_NAME, source_suffix)):
+    source_file_nosuf, _ = os.path.splitext(os.path.basename(source_file))
+    extensions.append(
+        Extension('{}.{}'.format(PACKAGE_NAME, source_file_nosuf),
+                  sources=['{}/{}{}'.format(
+                      PACKAGE_NAME, source_file_nosuf, source_suffix)],
+                  **extension_kwargs))
+
+if cythonize and 'clean' not in sys.argv:
+    extensions = cythonize(extensions, compiler_directives=cython_directives)
+
+
+class SDist(_sdist):
+    # Source distribution build runs Cython so that Cython is not needed as
+    # an install dependency.
+    def run(self):
+        cythonize(extensions, compiler_directives=cython_directives)
+        _sdist.run(self)
+
+
+def numpy_build_ext(pars):
```

Comment:
@tjtg This factory function pattern doesn't appear to work. 

After a bit of research, the source of this approach appears to come from https://stackoverflow.com/questions/19919905/how-to-bootstrap-numpy-installation-in-setup-py.

For me, I get the following:
```bash
> python setup.py build_ext --inplace
/data/local/itwl/miniconda3/envs/stratify/lib/python3.7/site-packages/setuptools/dist.py:472: UserWarning: Normalizing '0.2dev0' to '0.2.dev0'
  normalized_version,
Traceback (most recent call last):
  File "setup.py", line 113, in <module>
    setup(**setup_args)
  File "<snip>/lib/python3.7/site-packages/setuptools/__init__.py", line 145, in setup
    return distutils.core.setup(**attrs)
  File "<snip>/lib/python3.7/distutils/core.py", line 134, in setup
    ok = dist.parse_command_line()
  File "<snip>/lib/python3.7/site-packages/setuptools/dist.py", line 703, in parse_command_line
    result = _Distribution.parse_command_line(self)
  File "<snip>/lib/python3.7/distutils/dist.py", line 483, in parse_command_line
    args = self._parse_command_opts(parser, args)
  File "<snip>/lib/python3.7/site-packages/setuptools/dist.py", line 1018, in _parse_command_opts
    nargs = _Distribution._parse_command_opts(self, parser, args)
  File "<snip>/lib/python3.7/distutils/dist.py", line 545, in _parse_command_opts
    if not issubclass(cmd_class, Command):
TypeError: issubclass() arg 1 must be a class
```
`setuptools` is currently expecting a class (subclassed from `distutils.cmd.Command`) rather than a function. Perhaps this worked with a previous version of `setuptools`? The only way this *could* have worked is that `setuptools` was smart enough to call the factory function somehow in order to get the actual resultant class `NumpyBuildExt` - bet even then the factory function is returning an instantiation of the class, and not the class itself... so I'm finding that all rather confusing i.e. I can't see how the heck that would work (but perhaps I'm just being dense)

The point of using this factory function (highlighted in the cited stackoverflow) was to also delay the import of `from setuptools.command.build_ext import build_ext as _build_ext`, as `setuptools` attempts to import `Cython` itself from within its `setuptools.command.build_ext` module.

I noticed that you've moved the `from setuptools.command.build_ext import build_ext as _build_ext` out of the factory function (as per the cited stackoverflow) to be a module level import, so this is no longer a delayed import of `Cython` - and this is all because we wanted a delayed import of `numpy`.

That all said, is it now sufficient to just drop the use of the factory function, but keep the `NumpyBuildExt` class (unwrapped). Use the `NumpyBuildExt` class directly in the `cmdclass` along with a module level import of `from setuptools.command.build_ext import build_ext as _build_ext`, but also add `cython` is a dependency in the `setup_requires` and `install_requires`?

I tested `python setup.py build_ext --inplace` in a conda environment with only `setuptools` and `cython` installed and it seemed to work just fine.

Reply:
I tested this branch on some other external systems with varying python installation methods (distro packages, virtualenvs, conda) yesterday afternoon and ran into a bunch of problems that I hadn't encountered before, including similar to this one.

Adding cython as a dependency (rather than trying to avoid it as I was doing in this pull request) will simplify the code, and it's likely that users of this library will have cython installed for other purposes.

This is going to need some further reading, testing and probably some re-writing, so I'll close this pull request and open another with a fresh branch once I've got it into a state that's more compatible with a variety of systems/versions.

@bjlittle Thanks for your help with reviewing and testing.